drivers/serial/uart_nrfx_uarte.c

```diff
@@ -0,0 +1,613 @@
+		 * otherwise such operation may result in HardFault or RAM
+		 * corruption.
+		 */
+		nrf_uarte_tx_buffer_set(uarte, data->tx_buffer, 0);
+		nrf_uarte_task_trigger(uarte, NRF_UARTE_TASK_STARTTX);
+
+		/* switch off transmitter to save an energy */
+		nrf_uarte_task_trigger(uarte, NRF_UARTE_TASK_STOPTX);
+	}
+#endif
+
+	return 0;
+}
+
+#define UART_NRF_UARTE_DEVICE(idx)					      \
+	DEVICE_DECLARE(uart_nrfx_uarte##idx);				      \
+	UARTE_##idx##_CONFIG_INIT();					      \
+	static int uarte_##idx##_init(struct device *dev)		      \
+	{								      \
+		const struct uarte_init_config init_config = {		      \
+			.pseltxd = CONFIG_UART_##idx##_NRF_TX_PIN,	      \
+			.pselrxd = CONFIG_UART_##idx##_NRF_RX_PIN,	      \
+			UARTE_##idx##_NRF_HWFC_CONFIG			      \
+			.parity = UARTE_##idx##_NRF_PARITY_BIT,		      \
+			.baudrate = CONFIG_UART_##idx##_BAUD_RATE	      \
+		};							      \
+		UARTE_##idx##_INTERRUPTS_INIT();			      \
+		return uarte_instance_init(dev,				      \
+				&init_config,				      \
+				UARTE_##idx##_INTERRUPT_DRIVEN);	      \
+	}								      \
+	DEVICE_AND_API_INIT(uart_nrfx_uarte##idx,			      \
+			    CONFIG_UART_##idx##_NAME,                         \
+			    uarte_##idx##_init,                               \
+			    &uarte_##idx##_data,                              \
+			    &uarte_##idx##_config,                            \
+			    PRE_KERNEL_1,                                     \
+			    CONFIG_KERNEL_INIT_PRIORITY_DEVICE,               \
+			    &uart_nrfx_uarte_driver_api)
+
+#define UARTE_NRF_HWFC_ENABLED(idx)			\
+	.pselcts = CONFIG_UART_##idx##_NRF_CTS_PIN,	\
+	.pselrts = CONFIG_UART_##idx##_NRF_RTS_PIN,	\
+	.hwfc = NRF_UARTE_HWFC_ENABLED,
+#define UARTE_NRF_HWFC_DISABLED				\
+	.pselcts = NRF_UARTE_PSEL_DISCONNECTED,		\
+	.pselrts = NRF_UARTE_PSEL_DISCONNECTED,		\
+	.hwfc = NRF_UARTE_HWFC_DISABLED,
+
+#define UARTE_NRF_IRQ_ENABLED(idx)			\
+	IRQ_CONNECT(NRFX_IRQ_NUMBER_GET(NRF_UARTE##idx),\
+		    CONFIG_UART_##idx##_IRQ_PRI,        \
+		    uarte_nrfx_isr,                     \
+		    DEVICE_GET(uart_nrfx_uarte##idx),   \
+		    0)
+
+#define UARTE_TX_BUFFER_SIZE(idx)				\
+	CONFIG_UART_##idx##_NRF_TX_BUFFER_SIZE <		\
+		BIT_MASK(UARTE##idx##_EASYDMA_MAXCNT_SIZE) ?	\
+	CONFIG_UART_##idx##_NRF_TX_BUFFER_SIZE :		\
+		BIT_MASK(UARTE##idx##_EASYDMA_MAXCNT_SIZE)
+
+#define UARTE_CONFIG_INT(idx)						\
+	static u8_t uarte##idx##_tx_buffer[UARTE_TX_BUFFER_SIZE(idx)];	\
+	static struct uarte_nrfx_data uarte_##idx##_data = {		\
+		.tx_buffer = uarte##idx##_tx_buffer			\
+	};								\
+	static const struct uarte_nrfx_config uarte_##idx##_config = {	\
+		.uarte_regs = (NRF_UARTE_Type *)NRF_UARTE##idx,		\
+		.tx_buff_size = UARTE_TX_BUFFER_SIZE(idx)		\
+	}
+
+#define UARTE_CONFIG_NO_INT(idx)					\
+	static struct uarte_nrfx_data uarte_##idx##_data;		\
+	static const struct uarte_nrfx_config uarte_##idx##_config = {	\
+		.uarte_regs = (NRF_UARTE_Type *)NRF_UARTE##idx		\
```

Comment:
Can't you extract to macros only the parts that have to be different? The way it is done now, it is harder to maintain and more error-prone in case of modifications.

Reply:
done